draft-barnes-acme.md

```diff
@@ -1398,45 +1398,97 @@ If the server is able to validate the signature and confirm that the jwk and alg
+
+Once the client has generated the validation object and the MAC key, it serializes the validation object to UTF-8, and forms an HMAC-based JWS with the JSON object as its payload.  This JWS is NOT REQUIRED to have the "nonce" header parameter.
+
+The record provisioned to the DNS is the "signature" value from the JWS, i.e., the base64-encoded MAC value.  If so, the client constructs the validation domain name by appending the label "_acme-challenge" to the domain name being validated, then provisions a TXT record with the signature value under that name. For example, if the domain name being validated is "example.com", then the client would provision the following DNS record (with line breaks for display purposes only):
 
 ~~~~~~~~~~
 
-_acme-challenge.example.com. IN TXT "17817c66b60ce2e4012dfad92657527a"
+_acme-challenge.example.com. 300 IN TXT
+    "gfj9XqFv07e1wU66hSLYkiFqYakPSjAu8TsyXRg85nM"
 
 ~~~~~~~~~~
 
-The response to a DNS challenge is simply an acknowledgement that the relevant record has been provisioned.
+The response to a DNS challenge provides the ECDH key that the client generated and the HMAC-based JWS object.
 
 type (required, string):
 : The string "dns"
 
+clientPublicKey (required, JWK):
+: The ECDH key generated by the client
+
+validation (required, JWS):
+: The JWS object computed with the validation object and the derived key
+
 ~~~~~~~~~~
 
 {
   "type": "dns"
+  "clientPublicKey": { "kty": "EC", ... },
+  "validation": {
+    "header": { "alg": "HS256" },
+    "payload": "qzu9...6bjn",
+    "signature": "gfj9XqFv07e1wU66hSLYkiFqYakPSjAu8TsyXRg85nM"
+  }
 }
 
 ~~~~~~~~~~
 
-To validate a DNS challenge, the server queries for TXT records under the validation domain name.  If it receives a record whose contents match the token in the challenge, then the validation succeeds.  Otherwise, the validation fails.
+To validate a DNS challenge, the server performs the following steps:
+
+1. Derive a 256-bit MAC key using the "clientPublicKey" value and the private key corresponding to the "serverPublicKey" value
+2. Verify the "validation" JWS using the MAC key
+3. Decode the payload of the JWS as UTF-8 encoded JSON
+4. Verify that the there are exactly three fields in the decoded object, and that:
+  * The "type" field is set to "dns"
+  * The "token" field matches the "token" value in the challenge
+  * The "accountKey" field matches the account key for which the challenge was issued
+5. Query for TXT records under the validation domain name.
+6. Verify that the contents of one of the TXT records match the "signature" value in the "validation" JWS.
+
+If all of the above verifications succeed, then the validation is successful.  If no DNS record is found, or DNS record and response payload do not pass these checks, then the validation fails.
```

Comment:
The threat model is as described in the security considerations PR.  The value provisioned in the DNS record needs to be something that an ACME MitM can't use to authorize his own account key, assuming the client is honest.  Because the MAC key contains a component signed under the account key, the MitM can't force it to be something he knows (enabling him to spoof the DNS record), and because the MAC covers the account key, he can't use the DNS record value that the legitimate client produces.

That said, the only reason I proposed doing this with a MAC is that it didn't seem plausible to put a JWS in a DNS record.  But now that I look at it, we could put the signature value in the DNS record just as with the MAC value, and it wouldn't be appreciably bigger than, say, a DNSKEY record.  (Just the normal 33% base64 overhead.)

If that sounds better to you, I'll revise in the morning.